rfc-124-deploying-banners.md

```diff
@@ -0,0 +1,81 @@
+
+## Proposal
+
+Use Edge Side Include (ESI) to deploy banners on GOV.UK.
+
+Edge Side Include (ESI) is a [web standard](https://www.w3.org/TR/esi-lang) originally proposed by Akamai and Oracle, among other companies. It allows an Edge Server (like [Fastly’s caches](https://www.fastly.com/blog/using-esi-part-1-simple-edge-side-include)) to "mix and match" content from multiple URLs.
+
+In practice, the emergency banner on GOV.UK, for example, will require the following changes:
+
+1. make the emergency banner available as a fragment (e.g. on `/emergency-banner`); this should be hosted on a frontend server, potentially served by a new application called `banners`, similarly with the other public-facing application
+1. add an ESI include tag (pointing to the fragment) in the application layout
+```
+  <esi:include src="/emergency-banner" />
+```
+1. add an ESI remove tag and ESI comment to provide a fallback on environments where ESI is not available
+```
+  <esi:remove>
+    <%= render "/components/emergency_banner" %>
+  </esi:remove>
+  <!--esi
+  <esi:include src="/emergency-banner" />
+  -->
+```
+1. enable ESI in Fastly via Varnish Configuration Language (VCL)
+1. optional: add extra logic in VCL. Fastly gives us the ability to create custom responses at the edge – called "synthetics" in Varnish terminology – which can be used to include personalised, dynamic information that will be served from the edge instead of origin
+
+### Pros
+- performance gains (a higher cache-hit ratio, quicker page loads, less traffic spikes)
+- the include fragment will have its own URL and thus independent VCL logic and caching TTLs
+- due to the above, fragments can be purged independently of the main HTML content – this allows banners to roll-out instantly without having to drop the cache of all public-facing applications in GOV.UK
+- ESI remove can be used to store fallback HTML content (which could be the rendered banner component) and therefore make local development easier and more consistent with production
+
+### Cons
+- origin must serve top-level page HTML as an uncompressed response (or support content-encoding negotiation) in order for Varnish to be able to do streaming ESI replacement
+- it is also not possible for Varnish to compress the response after ESI processing. Fortunately Fastly has a workaround for this which forces H2O (Fastly’s HTTP terminator and HTTP/2-3 server) to do the streaming compression (which supports both `gzip` and `brotli`). This can easily be enabled via a HTTP response header which is subsequently stripped by H2O. However, as Fastly bills from bytes out of Varnish this would result in you getting billed for the uncompressed bytes instead of the compressed version. We are on a flat fee plan and thus this should not be an issue
+- if ESI is not implemented correctly can make us vulnerable to different attacks (SSRF and XSS)
```

Comment:
```suggestion
- if ESI is not implemented correctly, it could make us vulnerable to different attacks (SSRF and XSS)
```

Could you elaborate on this?

Reply:
_If_ an attacker manages to add an ESI include tag to the HTTP response, through publishing platforms or (reflected) XSS, they can perform SSRF attacks in the context of a surrogate server. Of course, we have cyber to test these things, but it's worth bearing in mind so we close as many doors as possible from the start.